bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/DefaultSWTFontRegistry.java

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Yatta Solutions
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Yatta Solutions - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.swt.internal;
+
+import java.util.*;
+
+import org.eclipse.swt.graphics.*;
+import org.eclipse.swt.internal.win32.*;
+
+/**
+ * This class is used in the win32 implementation only to support
+ * scaling of fonts in multiple DPI zoom level.
+ */
+public final class DefaultSWTFontRegistry implements SWTFontRegistry {
+	private static FontData KEY_SYSTEM_FONTS = new FontData();
+	private Map<FontData, Font> fontsMap = new HashMap<>();
+	private Device device;
+
+	public DefaultSWTFontRegistry(Device device) {
```

Comment:
Is there a convention for that? I didn't find that at any class in the internal package. But I don't mind adding it.

Reply:
I am not sure what the semantics of a `@noreference` within an internal package should be. Clients cannot reference that class/method anyway.

Reply:
That's true, I was aiming for something else though: no one (not even internal code) should instantiate this class without going through the factory-method `org.eclipse.swt.internal.SWTFontProvider.newFontRegistry(Device)`. One exception to this rule are the tests (`org.eclipse.swt.internal.DefaultSWTFontRegistryTests`).

You could achieve this by:
1. Make the constructor package-protected 
2. (Optionally) Mark the **class** with `@noinstantiate` or mark the constructor with `@noreference`. In both cases you should mention the factory-method in the JavaDoc

If point nr. 2. is an overkill, at least 1. should be used.

Reply:
> > 1. Making the constructor package-protected
> 
> This is basically reasonable, but due to the lack of possibility to write proper unit tests by test fragments, making the constructor `public` is is the trade-off to achieve testability here.

Good point: reducing the visibility lets the code still compile but the tests fail (at least locally they do) 👍 



> This would only make sense as a "workaround" for proposal 1 not being possible. But in general, even though it does not make sense to instantiate the registry, I do not see a reason from a design/architectural perspective to disallow the instantiation.

The reason would be "Defensive programming": if you don't need it, don't offer it. If you offer it, you have to maintain it. Also, the registry has to be disposed at some point to avoid resource leaking. The `SWTFontProvider` takes care of that. If this is going to be left for the user-code to do then at least it should be mentioned that the `dispose` method needs to be called when instantiating the class directly.

Reply:
Anything left to do here?